### PR TITLE
Implement more of GameEngine::Init

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/common/system/buildassistant.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        Object *>(
+        PICK_ADDRESS(0x004B5110, 0x009FA87E), this, what, angle, start, end, tiling_size, max_tiles, constructor_object);
+#else
+    return nullptr;
+#endif
+}
+
+Coord3D *BuildAssistant::Get_Build_Locations()
+{
+    return m_buildPositions;
```
Comment (xezon) [inline src/game/common/system/buildassistant.cpp:160]: Names are correct? `Get_Build_Locations` and `m_buildPositions` ?

File: src/game/logic/ai/aipathfind.cpp [hunk 2]
```diff
@@ -16,6 +16,220 @@
...
+
+        while (m_firstFree != nullptr) {
+            count++;
+            captainslog_dbgassert(m_firstFree->m_isFree, "Should be freed.");
+            m_firstFree = m_firstFree->m_pathParent;
+        }
+
+        captainslog_dbgassert(count == 30000, "Error - Allocated cellinfos.");
+        delete[] m_infoArray;
+        m_infoArray = nullptr;
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.cpp:102]: `m_cellInfos` perhaps? Judging by that assert message above.

File: src/game/logic/ai/aipathfind.cpp [hunk 3]
```diff
@@ -16,6 +16,220 @@
...
+
+void PathfindCellInfo::Release_Cell_Infos()
+{
+    if (m_infoArray != nullptr) {
+        int count = 0;
+
+        while (m_firstFree != nullptr) {
+            count++;
+            captainslog_dbgassert(m_firstFree->m_isFree, "Should be freed.");
+            m_firstFree = m_firstFree->m_pathParent;
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.cpp:97]: Is this loop supposed to do nothing other than asserting?

File: src/game/logic/ai/aipathfind.cpp [hunk 4]
```diff
@@ -16,6 +16,220 @@
...
+#ifdef GAME_DLL
+    Call_Method<void, PathfindLayer>(PICK_ADDRESS(0x0055DCA0, 0x0088E85F), this);
+#endif
+}
+
+#ifndef GAME_DLL
+PathfindCellInfo *PathfindCellInfo::m_infoArray;
+PathfindCellInfo *PathfindCellInfo::m_firstFree;
+#endif
+
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.cpp:74]: constexpr int max_cell_info_count = 30000

File: src/game/logic/ai/aipathfind.cpp [hunk 5]
```diff
@@ -16,6 +16,220 @@
...
+void PathfindCellInfo::Allocate_Cell_Infos()
+{
+    Release_Cell_Infos();
+    m_infoArray = new PathfindCellInfo[30000];
+    m_infoArray[29999].m_pathParent = 0;
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.cpp]: Add a global `constexpr int max_cell_count = 30000` in this file and then use it in these functions, like `max_cell_count - 1` here.

File: src/game/logic/ai/ai.h [hunk 6]
```diff
@@ -58,6 +83,7 @@ class AISideBuildList : public MemoryPoolObject
     Utf8String m_name;
     BuildListInfo *m_buildListInfo;
     AISideBuildList *m_next;
+    friend class TAiData;
 };
 
 class TAiData : public SnapShot
```
Comment (xezon) [inline src/game/logic/ai/ai.h:89]: `TAiData` is intended name? Not `AIData`?

File: src/game/logic/ai/ai.h [hunk 7]
```diff
@@ -116,14 +142,16 @@ class TAiData : public SnapShot
...
-    AISideInfo *m_SideInfo;
+    AISideInfo *m_sideInfo;
     AISideBuildList *m_AISideBuildList;
     TAiData *m_next;
 };
 
 class AI : public SubsystemInterface, public SnapShot
 {
 public:
+    AI *Hook_Ctor() { return new (this) AI(); }
```
Comment (xezon) [inline src/game/logic/ai/ai.h:154]: `#ifdef GAME_DLL` missing?

File: src/game/logic/ai/ai.cpp [hunk 8]
```diff
@@ -24,6 +117,59 @@ void AIGroup::Add(Object *obj)
...
+    m_pathfinder = new Pathfinder();
+}
+
+AI::~AI()
+{
+    if (m_pathfinder != nullptr) {
+        delete m_pathfinder;
+    }
+
+    m_pathfinder = nullptr;
```
Comment (xezon) [inline src/game/logic/ai/ai.cpp:132]: Can go into scope above, or removed, because this is destructor.

File: src/game/logic/ai/ai.cpp [hunk 9]
```diff
@@ -24,6 +117,59 @@ void AIGroup::Add(Object *obj)
 #endif
 }
 
+AI::AI() : m_formationID(0)
```
Comment (xezon) [inline src/game/logic/ai/ai.cpp]: `int m_groupID` not initialized.

File: src/game/logic/ai/aipathfind.cpp [hunk 10]
```diff
@@ -16,6 +16,220 @@
...
+}
+
+PathfindZoneManager::~PathfindZoneManager()
+{
+#ifdef GAME_DLL
+    Call_Method<void, PathfindZoneManager>(PICK_ADDRESS(0x0055C320, 0x0088BF4E), this);
+#endif
+}
+
+PathfindLayer::PathfindLayer() : m_map(nullptr), m_mapPointers(nullptr), m_unk(0), m_bridge(nullptr), m_destroyed(false)
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.cpp]: `PathfindLayerEnum m_layerNum` not initialized.

File: src/game/logic/ai/aipathfind.h [hunk 11]
```diff
@@ -57,13 +82,154 @@ class Path : public MemoryPoolObject, public SnapShot
...
+        const LocomotorSet &locomotor_set,
+        const Coord3D *pos1,
+        const Coord3D *pos2,
+        const Coord3D *pos3,
+        float f) = 0;
+};
+
+class PathfindCellInfo
 {
 public:
```
Comment (xezon) [inline src/game/logic/ai/aipathfind.h:131]: Constructor will be added later?

File: src/game/common/system/upgrade.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    ACT_NONE,
+    ACT_UPGRADE_RADAR,
+    ACT_SUPERPOWER,
+};
+
+class UpgradeTemplate : public MemoryPoolObject
+{
+    IMPLEMENT_POOL(UpgradeTemplate);
+
+public:
```
Comment (xezon) [inline src/game/common/system/upgrade.h:41]: Constructor will be done later?

File: src/game/common/system/buildassistant.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/common/system/buildassistant.cpp:160]: Yep, both names are correct.

File: src/game/logic/ai/aipathfind.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/logic/ai/aipathfind.cpp:102]: m_infoArray is correct.

File: src/game/logic/ai/aipathfind.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/ai/aipathfind.cpp:97]: Yes, it's supposed to just walk through setting m_firstFree to m_firstFree->m_pathParent.

File: src/game/logic/ai/aipathfind.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/ai/aipathfind.cpp:74]: fixed.

File: src/game/logic/ai/ai.h (hunk 6 above)
Comment (jonwil) [reply src/game/logic/ai/ai.h:89]: This is correct.

File: src/game/logic/ai/ai.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/ai/ai.cpp]: Its initialized elsewhere.

File: src/game/logic/ai/aipathfind.h (hunk 11 above)
Comment (jonwil) [reply src/game/logic/ai/aipathfind.h:131]: There is no constructor for PathfindCellInfo at all (none exists even where you would expect one in ZH WB)

File: src/game/common/system/upgrade.h (hunk 12 above)
Comment (jonwil) [reply src/game/common/system/upgrade.h:41]: Constructor for this will be done later when its needed.

File: src/game/logic/ai/ai.h (hunk 7 above)
Comment (jonwil) [reply src/game/logic/ai/ai.h:154]: Fixed.

File: src/game/logic/ai/aipathfind.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/ai/aipathfind.cpp]: fixed.

**Comment by xezon** [comment]
Shellmap looks normal, Building, Producing, Selling looks normal, Unit & Structure upgrades work, Enemy AI and Pathfinding looks as usual. Big test map with all units survives stress test. No crashes. No obvious glitches.

Tested in RelWithDebInfo.

**Review by xezon — APPROVED** [review]

**Pull request merged**